tools/saftw.py

```diff
@@ -0,0 +1,310 @@
+#!/usr/bin/python
+#
+from ftw import ruleset, util, testrunner, http
+import getopt
+import os
+import sys
+import time
+import re
+
+class cmdopt(object):
+    rule_path = None
+    recurse = 0
+    journal = None
+    logfile = None
+    timeout = 1
+    replayjournal = 0
+    debug = 0
+    plugin = None
+    rulematch = 0
+    overload = {} 
+    nocache = 0
+
+def usage():
+    print "Produce journal:"
+    print "saftw -p <plugin> -j <journal file> -r <ftw rule/dir>"
+    print "Process journal:"
+    print "saftw -p <plugin> -J -l <logfile> -j <journal file>"
+    sys.exit(1)
+
+class waf_plugin(object):
+    def load(self, name):
+        if name == "varnish":
+            return varnish_waf()
+        if name == "modsec":
+            return modsec_waf()
+        raise
+
+    def get_log_tag(self, now, title):
+        raise NotImplementedError
+
+    def match(self, rule, ln):
+        raise NotImplementedError
+
+class modsec_waf(waf_plugin):
+    name = "modsec"
+
+    def __init__(self):
+        try:
+            self.prog = re.compile('^.* \[id "(\d+)"')
+        except:
+            print "failed to compile regexp"
+            sys.exit(1)
+        super(modsec_waf, self).__init__()
+
+    def get_log_tag(self, now, title):
+        t = time.gmtime(now)
+        ts = time.strftime("%a %b %d %H:%M:%S", t)
+        '''
+            NB: we need to return a list here on the off
+            chance we are on a boundary
+        '''
+        return ts
+
+    def match(self, rule, ln):
+        m = self.prog.match(ln)
+        if m == None:
+            return False
+        if int(m.group(1)) == rule:
+            return True
+        return False
+
+class varnish_waf(waf_plugin):
```

Comment:
We should probably get rid of this, since the logging format would be user defined anyway